File: python/timesift/registry.py
"""One registry mechanism, used by the three things the package is meant to be extended with.

Learners, response heads and metrics all register the same way, so adding any of them is a
registration and never an edit to the code that fits or scores: the fitting path holds no list of
names. The R package carries the same three registries under the same names.
"""

from __future__ import annotations

from typing import Callable


class Registry:
    """A named collection of one kind of extension."""

    def __init__(self, what: str):
        self.what = what
        self._entries: dict[str, object] = {}

    def set(self, name: str, value, overwrite: bool = False):
        if not isinstance(name, str) or not name:
            raise ValueError(f"a {self.what} needs a single non-empty name")
        if name in self._entries and not overwrite:
            raise ValueError(f'{self.what} "{name}" is already registered. '
                             "Pass overwrite=True to replace it.")
        self._entries[name] = value
        return value

    def get(self, name: str):
        if name not in self._entries:
            raise KeyError(f'unknown {self.what} "{name}". Registered: '
                           f'{", ".join(self.names())}.')
        return self._entries[name]

    def has(self, name: str) -> bool:
        return name in self._entries

    def remove(self, name: str) -> None:
        self._entries.pop(name, None)

    def names(self) -> list[str]:
        # C collation, so a registry lists the same order on every machine and beside R's.
        return sorted(self._entries)


LEARNERS = Registry("learner")
RESPONSES = Registry("response")
METRICS = Registry("metric")


def resolve_metric(metric, default: str | None = None):
    """The function that scores and the name a report prints, from either way a metric is given.

    A metric reaches a run as a registered name or as a function of ``(y, p)``, and everything
    that rescores afterwards -- the ensemble row of a report, an occlusion profile -- needs the
    function rather than a name to look up again. Both travel with the fit. A function has no name
    to print, and reads as ``<function>`` on both sides rather than as whatever the language calls
    an anonymous one.
    """
    if metric is None:
        metric = default
    if callable(metric):
        return metric, "<function>"
    if not isinstance(metric, str):
        raise TypeError("`metric` is the name of a registered metric or a function of (y, p), "
                        f"got {type(metric).__name__}")
    return METRICS.get(metric), metric


def register_learner(name: str, constructor: Callable, overwrite: bool = False):
    """Make a learner available by name. The learners that ship are registered the same way.

    ``constructor`` is called with no arguments and returns a ``Learner``, so a learner asked for
    by name is built with its own defaults.
    """
    if not callable(constructor):
        raise ValueError("a learner is registered as a constructor taking no arguments")
    return LEARNERS.set(name, constructor, overwrite)


def learners() -> list[str]:
    """The learners registered under this session."""
    return LEARNERS.names()


def get_learner(learner):
    """A ``Learner``, whether it arrived as one or as the name of a registered one."""
    from .learners import Learner
    if isinstance(learner, Learner):
        return learner
    return LEARNERS.get(learner)()


def register_metric(name: str, fn: Callable, overwrite: bool = False):
    """Register a metric: a function of ``(y, p)`` on one held-out cell.

    ``y`` and ``p`` are the observed values of the units in one fold and a model's predictions for
    them, both the same length. It returns one number, or a value that is not finite where the cell
    defines none. Registering one makes it available to ``grain_ladder`` by name, with no change
    to the fitting code.
    """
    if not callable(fn):
        raise ValueError("a metric is a function of (y, p)")
    return METRICS.set(name, fn, overwrite)


def metrics() -> list[str]:
    """The metrics registered under this session."""
    return METRICS.names()


RESPONSE_FIELDS = ("prepare", "activation", "loss", "metric", "cells")


def register_response(name: str, spec: dict, overwrite: bool = False):
    """Register a response head: what the values being predicted are and where a score is defined.

    ``spec`` is a mapping with ``prepare(y)``, returning the response a learner is fitted on;
    ``activation``, the name of the output transform (``"sigmoid"`` or ``"identity"``); ``loss``,
    the name of the training objective (``"binary_cross_entropy"`` or ``"squared_error"``);
    ``metric``, the default metric name; and ``cells(y, folds)``, returning the mask of scorable
    cells. Presence-absence with a joint multi-label head is what ships; an abundance or
    phenology response is a registration rather than a second fitting path. Every learner that
    ships reads ``loss`` and ``activation`` from here: the encoders train under the loss and
    predict through the activation, and the learners fitting one model per response take the
    family the loss names, logistic or Gaussian. The combiner minimises the same loss.
    """
    missing = [f for f in RESPONSE_FIELDS if f not in spec]
    if missing:
        raise ValueError(f"a response specification needs {', '.join(missing)}")
    return RESPONSES.set(name, spec, overwrite)


def responses() -> list[str]:
    """The response heads registered under this session."""
    return RESPONSES.names()

File: R/registry.R
# One registry mechanism, used by the three things the package is meant to be extended with:
# learners, response heads and metrics. Adding any of them is a registration, never an edit to the
# code that fits or scores, so the fitting path has no list of names in it.

.new_registry <- function(what) {
  entries <- new.env(parent = emptyenv())
  list(
    set = function(name, value, overwrite = FALSE) {
      if (!is.character(name) || length(name) != 1L || !nzchar(name)) {
        stop("a ", what, " needs a single non-empty name.", call. = FALSE)
      }
      if (!overwrite && exists(name, envir = entries, inherits = FALSE)) {
        stop(what, " \"", name, "\" is already registered. ",
             "Pass `overwrite = TRUE` to replace it.", call. = FALSE)
      }
      assign(name, value, envir = entries)
      invisible(value)
    },
    get = function(name) {
      if (!exists(name, envir = entries, inherits = FALSE)) {
        stop("unknown ", what, " \"", name, "\". Registered: ",
             paste(sort(ls(entries), method = "radix"), collapse = ", "), ".", call. = FALSE)
      }
      get(name, envir = entries, inherits = FALSE)
    },
    has = function(name) exists(name, envir = entries, inherits = FALSE),
    remove = function(name) {
      if (exists(name, envir = entries, inherits = FALSE)) rm(list = name, envir = entries)
      invisible(NULL)
    },
    # C collation, so a registry lists the same order on every machine and beside Python's.
    names = function() sort(ls(entries), method = "radix")
  )
}

.learners_reg <- .new_registry("learner")
.responses_reg <- .new_registry("response")
.metrics_reg <- .new_registry("metric")

#' Register a metric
#'
#' A metric scores one held-out cell: the observed response of the units in a fold and a model's
#' predictions for them. Registering one makes it available to [grain_ladder()] by name, with no
#' change to the fitting code.
#'
#' @param name Name the metric is asked for by.
#' @param fn A function of `(y, p)`, the observed values and the predictions for one cell, both
#'   vectors of the same length, returning one number or `NA` where the cell defines none.
#' @param overwrite Replace an existing registration.
#'
#' @return The registered function, invisibly.
#'
#' @examples
#' register_metric("hit_rate", function(y, p) mean((p >= 0.5) == (y == 1)), overwrite = TRUE)
#' "hit_rate" %in% metrics()
#'
#' @export
register_metric <- function(name, fn, overwrite = FALSE) {
  if (!is.function(fn)) {
    stop("a metric is a function of (y, p).", call. = FALSE)
  }
  .metrics_reg$set(name, fn, overwrite)
}

#' @rdname register_metric
#' @export
metrics <- function() .metrics_reg$names()

# A metric reaches a run either as a registered name or as a function of (y, p), and everything
# that rescores afterwards -- the ensemble row of a report, an occlusion profile -- needs the
# function rather than a name to look up again. Both travel with the fit: the function is what
# scores, the name is what the report prints. A function has no name to print, and reads as
# `<function>` on both sides rather than as whatever the language calls an anonymous one.
.as_metric <- function(metric, default = NULL) {
  metric <- metric %||% default
  if (is.function(metric)) {
    return(list(fn = metric, name = "<function>"))
  }
  if (!is.character(metric) || length(metric) != 1L) {
    stop("`metric` is the name of a registered metric or a function of (y, p), got ",
         class(metric)[1L], ".", call. = FALSE)
  }
  list(fn = .metrics_reg$get(metric), name = metric)
}

#' Register a response head
#'
#' A response head says what the values being predicted are, how they reach a learner, and which
#' cells of the (variable, fold) grid a score is defined on. Presence-absence with a joint
#' multi-label head is what ships; an abundance or phenology response is a registration rather than
#' a second fitting path.
#'
#' @param name Name the response is asked for by.
#' @param spec A list with elements `prepare(y)`, returning the numeric matrix a learner is fitted
#'   on; `activation`, the name of the output transform (`"sigmoid"` or `"identity"`); `loss`, the
#'   name of the training objective (`"binary_cross_entropy"` or `"squared_error"`); `metric`, the
#'   default metric name; and `cells(y, folds)`, returning the mask of scorable cells. Every
#'   learner that ships reads `loss` and `activation` from here: the encoders train under the loss
#'   and predict through the activation, and the learners fitting one model per response take the
#'   family the loss names, logistic or Gaussian. The combiner minimises the same loss.
#' @param overwrite Replace an existing registration.
#'
#' @return The registered specification, invisibly.
#'
#' @examples
#' responses()
#'
#' @export
register_response <- function(name, spec, overwrite = FALSE) {
  need <- c("prepare", "activation", "loss", "metric", "cells")
  missing <- setdiff(need, names(spec))
  if (length(missing)) {
    stop("a response specification needs ", paste(missing, collapse = ", "), ".", call. = FALSE)
  }
  .responses_reg$set(name, spec, overwrite)
}

#' @rdname register_response
#' @export
responses <- function() .responses_reg$names()
